engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -0,0 +1,152 @@
+package cromwell.engine.workflow
+
+import java.time.OffsetDateTime
+import java.util.UUID
+
+import akka.actor.{Actor, Props}
+import akka.event.Logging
+import cromwell.core.{ErrorOr, WorkflowId, WorkflowSourceFiles}
+import cromwell.database.obj.WorkflowMetadataKeys
+import cromwell.engine.workflow.WorkflowStore.WorkflowToStart
+import cromwell.engine.workflow.WorkflowStoreActor._
+import cromwell.services.MetadataServiceActor.PutMetadataAction
+import cromwell.services.{MetadataEvent, MetadataKey, MetadataValue, ServiceRegistryClient}
+
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+import scalaz.NonEmptyList
+
+object WorkflowStore {
+  sealed trait WorkflowState {
+    def isStartable: Boolean = {
+      this match {
+        case x: StartableState => true
+        case _ => false
+      }
+    }
+  }
+
+  case object Running extends WorkflowState
+
+  sealed trait StartableState extends WorkflowState
+  case object Submitted extends StartableState
+  case object Restartable extends StartableState
+
+  final case class SubmittedWorkflow(id: WorkflowId, sources: WorkflowSourceFiles, state: WorkflowState) {
+    def toWorkflowToStart: WorkflowToStart = {
+      state match {
+        case r: StartableState => WorkflowToStart(id, sources, r)
+        case _ => throw new IllegalArgumentException("This workflow is not currently in a startable state")
+      }
+    }
+  }
+
+  final case class WorkflowToStart(id: WorkflowId, sources: WorkflowSourceFiles, state: StartableState)
+}
+
+/**
+  * Stores any submitted workflows which have not completed.
+  *
+  * This is not thread-safe, although it should not be being used in a multithreaded fashion. If that
+  * changes, revisit.
+  */
+abstract class WorkflowStore {
+  import WorkflowStore._
+
+  var workflowStore = List.empty[SubmittedWorkflow]
+
+  /**
+    * Adds the requested WorkflowSourceFiles to the store and returns a WorkflowId for each one (in order)
+    * for tracking purposes.
+    */
+  def add(sources: NonEmptyList[WorkflowSourceFiles]): NonEmptyList[WorkflowId] = {
+    val submittedWorkflows = sources map { SubmittedWorkflow(WorkflowId.randomId(), _, Submitted) }
+    workflowStore = workflowStore ++ submittedWorkflows.list
+    submittedWorkflows map { _.id }
+  }
+
+  /**
+    * Retrieves up to n workflows which have not already been pulled into the engine and sets their pickedUp
+    * flag to true
+    */
+  def fetchRunnableWorkflows(n: Int): List[WorkflowToStart] = {
+    val startableWorkflows = workflowStore filter { _.state.isStartable } take n
+    val updatedWorkflows = startableWorkflows map { _.copy(state = Running) }
+    workflowStore = (workflowStore diff startableWorkflows) ++ updatedWorkflows
+
+    startableWorkflows map { _.toWorkflowToStart }
+  }
+
+  def remove(id: WorkflowId): Option[Unit] = {
+    val newWorkflowStore = workflowStore filterNot { _.id == id }
+
```

Comment:
Looking at this again I'm even less in love w/ it vs when i first did it. But this portion doesn't have access to the logger, yet using a Try seemed pointless since the akka logger doesn't take exceptions. Open to suggestions